webapp-django/crashstats/api/views.py

```diff
@@ -51,72 +50,50 @@
 )
 
 
-class APIAllowlistError(Exception):
-    pass
+class MultipleStringField(forms.TypedMultipleChoiceField):
+    """
+    Field that does not validate if the field values are in self.choices
 
+    Validators are run on each item in the list, rather than against the whole input,
+    like other Django fields.
+    """
+    default_validators = [ProhibitNullCharactersValidator()]
 
-class MultipleStringField(forms.TypedMultipleChoiceField):
-    """Field that do not validate if the field values are in self.choices"""
+    def valid_value(self, value):
+        """
+        Run validators on each item in the list.
 
-    def validate(self, value):
-        """Nothing to do here"""
-        if self.required and not value:
-            raise forms.ValidationError(self.error_messages['required'])
+        The TypedMultipleChoiceField.valid_value method checks that
+        the string is in self.choices, and this version explictly skips that check.
+        """
+        self.run_validators(value)
+        return True
 
 
 TYPE_MAP = {
     str: forms.CharField,
     list: MultipleStringField,
     datetime.date: forms.DateField,
+    # NOTE: Not used in any API models
     datetime.datetime: forms.DateTimeField,
     int: forms.IntegerField,
-    bool: forms.BooleanField,
+    # NOTE: Not used in any API models
+    bool: forms.NullBooleanField,
```

Comment:
[NullBooleanField](https://docs.djangoproject.com/en/2.2/_modules/django/forms/fields/#NullBooleanField) does what this code was doing in the ``clean`` method:

```
        if value in (True, 'True', 'true', '1'):
            return True
        elif value in (False, 'False', 'false', '0'):
            return False
        else:
            return None
```

Switching to ``NullBooleanField`` simplifies the ``MiddlewareModelForm`` class, which can drop the custom ``clean`` method entirely.

The docs for ``BooleanField`` suggest using ``null=True`` instead, because they are threatening to deprecate ``NullBooleanField`` in the future. I'm inclined to wait until they actually add a deprecation warning, which isn't in the Django 3.0 / master code yet. Or, since no API models use ``bool`` or ``datetime``, drop those two types from this list.

Reply:
+1 for switching to `NullBooleanField`.

If you're at all concerned about `NullBooleanField` getting nixed, we could write our own that subclasses `BooleanField` and passes `null=True` into `__init__`. That seems pretty straight-forward.

Does Supersearch API request handling go through code that uses TYPE_MAP? If so, I'm pretty sure we use bool and datetime there.

Reply:
You are right, our own ``NullBooleanField`` would be a nice solution if the Django one goes away. I still expect Django 4.0 at the earliest.

The ``TYPE_MAP`` is only used by this code. The ``get_annotated_params`` code that generates the type names only supports ``str``, ``int``, and ``tuple``:

https://github.com/mozilla-services/socorro/blob/a6792a9ed8e9da42758eb4f39e6677730614a9a8/webapp-django/crashstats/crashstats/models.py#L571-L597

The related ``kwargs_to_params`` is more flexible, handling ``datetime`` and ``bool``:

https://github.com/mozilla-services/socorro/blob/a6792a9ed8e9da42758eb4f39e6677730614a9a8/webapp-django/crashstats/crashstats/models.py#L507-L569

There's some similar code, a ``TYPE_TO_FIELD_MAPPING`` that includes ``bool`` but not ``datetime``. I believe these are also dead code paths:

https://github.com/mozilla-services/socorro/blob/a6792a9ed8e9da42758eb4f39e6677730614a9a8/webapp-django/crashstats/supersearch/forms.py#L12-L19